api/src/main/java/org/openmrs/layout/name/NameSupport.java

```diff
@@ -27,16 +38,100 @@ public NameSupport() {
-		if (singleton == null) {
-			throw new APIException("Not Yet Instantiated");
-		} else {
-			return singleton;
+		synchronized (NameSupport.class) {
+			singleton.init();
+			if (singleton == null) {
+				throw new APIException("Not Yet Instantiated");
+			} else {
+				return singleton;
+			}
 		}
 	}
+
+	/**
+	 * Initializes layout templates with a custom template configured
+	 * via the "layout.name.template" GP.
+	 */
+	private void init() {
+		if (initialized) {
+			return;
+		}
+		Context.getAdministrationService().addGlobalPropertyListener(singleton);
+		// Get configured name template to override the existing one if any
+		String layoutTemplateXml = Context.getAdministrationService().getGlobalProperty(
+			OpenmrsConstants.GLOBAL_PROPERTY_LAYOUT_NAME_TEMPLATE);
+		NameTemplate nameTemplate = deserializeXmlTemplate(layoutTemplateXml);
+		
+		if (nameTemplate != null) {
+			updateLayoutTemplates(nameTemplate);
+			initialized = true;
+		}
+	}
+	
+	/**
+	 * Update existing layout templates if present with the provided template
+	 */
+	private void updateLayoutTemplates(NameTemplate nameTemplate) {
+		if (getLayoutTemplates() == null) {
+			setLayoutTemplates(new ArrayList<>());
+		}
+		List<NameTemplate> list = new ArrayList<>();
+		// filter out unaffected templates to keep
+		list.addAll(getLayoutTemplates().stream().filter(existingTemplate -> existingTemplate.getCodeName() != nameTemplate.getCodeName()).collect(Collectors.toList()));
+		list.add(nameTemplate);
+		setLayoutTemplates(list);
+	}
 	
+	/**
+	 * @return Returns the defaultLayoutFormat
+	 */
+	private NameTemplate deserializeXmlTemplate(String xml) {
+		NameTemplate nameTemplate = null;
+		try {
+			nameTemplate = Context.getSerializationService().getDefaultSerializer().deserialize(xml,
+				NameTemplate.class);
+		} catch (Exception e) {
+			log.error("Error in deserializing provided name template", e);
+		}
+		return nameTemplate;
+	}
+
+	/**
+	 * @return Returns the defaultLayoutFormat
+	 */
 	@Override
 	public String getDefaultLayoutFormat() {
-		String ret = Context.getAdministrationService().getGlobalProperty("layout.name.format");
+		String ret = Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_LAYOUT_NAME_FORMAT);
 		return (ret != null && ret.length() > 0) ? ret : defaultLayoutFormat;
 	}
+
+	/**
+	 * @see org.openmrs.api.GlobalPropertyListener#supportsPropertyName(String)
+	 */
+	@Override
+	public boolean supportsPropertyName(String propertyName) {
+		return OpenmrsConstants.GLOBAL_PROPERTY_LAYOUT_NAME_TEMPLATE.equals(propertyName);
+	}
+
+	/**
+	 * @see org.openmrs.api.GlobalPropertyListener#globalPropertyChanged(org.openmrs.GlobalProperty)
+	 */
+	@Override
+	public void globalPropertyChanged(GlobalProperty newValue) {
+		if (!OpenmrsConstants.GLOBAL_PROPERTY_LAYOUT_NAME_TEMPLATE.equals(newValue.getPropertyValue())) {
+			return;
+		}
+		NameTemplate nameTemplate = deserializeXmlTemplate(newValue.getPropertyValue());
```

Comment:
What happens when one clears out the global property value? Don't you then want to also clear the cached template?

Reply:
The cached translate then cannot be fetched/queried. It'd stay until the next restart. This is related to the current  implementation of supported `NameTemplate`s. Removing one requires a restart.